### PR TITLE
Fix Validation hole bug

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: include/veriblock/blockchain/block_index.hpp [hunk 1]
```diff
@@ -19,7 +19,7 @@
 
 namespace altintegration {
 
-enum BlockStatus : uint8_t {
+enum BlockStatus : uint16_t {
```
Comment (Warchant) [inline include/veriblock/blockchain/block_index.hpp:22]:
```suggestion
enum BlockStatus : uint32_t {
```

File: include/veriblock/blockchain/pop/pop_state_machine.hpp [hunk 2]
```diff
@@ -14,6 +14,54 @@
 
 namespace altintegration {
 
+namespace {
+
+template <typename index_t>
+void applyBlockAsserts(index_t& index) {
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/pop_state_machine.hpp:20]:
```suggestion
void assertBlockCanBeApplied(index_t& index) {
```

File: include/veriblock/blockchain/pop/pop_state_machine.hpp [hunk 3]
```diff
@@ -14,6 +14,54 @@
...
+  VBK_ASSERT(!index.hasFlags(BLOCK_APPLIED) &&
+             "state corruption: tried to apply an already applied block");
+  // an expensive check; might want to  disable it eventually
+  VBK_ASSERT(index.allDescendantsUnapplied() &&
+             "state corruption: found an unapplied block that has some of "
+             "its descendants applied");
+}
+
+template <typename index_t>
+void unapplyBlockAsserts(index_t& index) {
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/pop_state_machine.hpp:43]:
```suggestion
void assertBlockCanBeUnapplied(index_t& index) {
```

File: include/veriblock/blockchain/block_index.hpp [hunk 4]
```diff
@@ -39,7 +39,13 @@ enum BlockStatus : uint8_t {
   BLOCK_FAILED_MASK =
       BLOCK_FAILED_CHILD | BLOCK_FAILED_POP | BLOCK_FAILED_BLOCK,
   //! the block has been applied via PopStateMachine
-  BLOCK_APPLIED = 1 << 5
+  BLOCK_APPLIED = 1 << 5,
+  //! the block has been at least once applied via PopStateMachine
+  BLOCK_ONCE_APPLIED = 1 << 6,
```
Comment (Warchant) [inline include/veriblock/blockchain/block_index.hpp:44]:
```suggestion
  BLOCK_CAN_BE_APPLIED = 1 << 6,
```

**Comment by Warchant** [comment]
Status is changed from uint8_t to uint32_t, which is a breaking change.

File: include/veriblock/blockchain/block_index.hpp [hunk 5]
```diff
@@ -39,7 +39,13 @@ enum BlockStatus : uint8_t {
...
+  //! the block has pop data
+  BLOCK_HAS_POP = 1 << 7,
```
Comment (Warchant) [inline include/veriblock/blockchain/block_index.hpp:46]: why do you need this flag? 
`index->hasPayloads` should work

File: include/veriblock/blockchain/pop/pop_state_machine.hpp (hunk 3 above)
Comment (Mr-Leshiy) [reply include/veriblock/blockchain/pop/pop_state_machine.hpp:38]: fixed

File: include/veriblock/blockchain/pop/pop_state_machine.hpp (hunk 2 above)
Comment (Mr-Leshiy) [reply include/veriblock/blockchain/pop/pop_state_machine.hpp:20]: fixed

File: include/veriblock/blockchain/block_index.hpp (hunk 1 above)
Comment (Mr-Leshiy) [reply include/veriblock/blockchain/block_index.hpp:22]: fixed

File: include/veriblock/blockchain/pop/pop_state_machine.hpp [hunk 6]
```diff
@@ -14,6 +14,48 @@
...
+  VBK_ASSERT_MSG(
+      index.allDescendantsUnapplied() && "state corruption: %s",
+      "found an unapplied block that has some of its descendants applied");
+}
+
+template <typename index_t>
+void assertBlockCanBeUnapplied(index_t& index) {
+  VBK_ASSERT(index.pprev && "cannot unapply the genesis block");
+
+  VBK_ASSERT_MSG(index.hasFlags(BLOCK_APPLIED) && "state corruption: %s",
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/pop_state_machine.hpp:41]:
```suggestion
  VBK_ASSERT_MSG(index.hasFlags(BLOCK_APPLIED), "state corruption: %s",
```

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 7]
```diff
@@ -552,7 +552,24 @@ struct PopAwareForkResolutionComparator {
     } else {
       // chain B is better. unapply A and leave B applied
       auto guard = ing_->deferForkResolutionGuard();
+      // if candidate has never been applied before the comparePopScore() so we
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/fork_resolution.hpp:555]:
```suggestion
      // if candidate has never been applied before the comparePopScore(), we
```

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 8]
```diff
@@ -552,7 +552,24 @@ struct PopAwareForkResolutionComparator {
...
+      // have carefully unapply this chain before unapplying chainA and trying
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/fork_resolution.hpp:556]:
```suggestion
      // have to unapply candidate before unapplying chainA and try
```

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 9]
```diff
@@ -552,7 +552,24 @@ struct PopAwareForkResolutionComparator {
...
+      }
       sm.unapply(*chainA.tip(), *chainA.first());
+
+      // validate chain
+      if (!canBeApplied && !sm.apply(*chainB.first(), *chainB.tip(), state)) {
+        bool res = sm.apply(*chainA.first(), *chainA.tip(), state);
+        VBK_ASSERT_MSG(res,
+                       "state corruption: %s",
+                       "chainA as a previous best chain should be valid");
+        VBK_LOG_INFO("Chain A wins");
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/fork_resolution.hpp:569]:
```suggestion
        VBK_LOG_INFO("Chain B is invalid when applied alone. Chain A wins");
```

**Comment by Warchant** [comment]
status is in-mem field, it looks like change is non-breaking

**Pull request merged**